feat: multiple SPA routers

edivados opened the pull request:
The goal of this PR is to allow the configuration of two or more SPA routers where each SPA is served from a differnt base.

# Current state

Currently a configuration as seen below does not work for the following reasons:

- Only one handler is generated instead of one per SPA router
- Vite project root is not set to the routers specified root
- Rollup resolving paths relative to the current working directory (Rollup has no option to change project root)
- Vite's `transformIndexHtml` function not transforming relative paths

```js
const app = createApp({
  routers: [
    {
      name: "public",
      mode: "static",
      dir: "./public",
    },
    {
      name: "react",
      mode: "spa",
      root: "./react",
      // You cannot choose the location of index.html Vite will look for it at the project root
     // .html ending is just an indicator to let Vite handle things
      handler: "./index.html",
      target: "browser",
      plugins: () => [react()]
    },
    {
      name: "solid",
      mode: "spa",
      root: "./solid",
      handler: "./index.ts",
      base: "/solid",
      target: "browser",
      plugins: () => [solid()]
    }
  ],
});
```

# Todos

- [x] Add test
- [x] Generate one handler per SPA router
- [x] Set Vite's project root to the router root
- [x] Transform relative paths in html served by handler in dev
- [x] Solve project root issue with Rollup
- [x] Public files not copied to output for routers with different root
- [x] Build warning about dir not being in project root (only for the solid router)
- [x] Test if `fallthrough` is applied correctly for public assets
- ~Multiple SPA routers sharing the same root (possible?)~

# Notes

<details>
<summary>Shared project root</summary>

How would one handle multiple `tsconfig.json`?

 If rollup input is `./solid/index.html` the output path will be `{outDir}/solid/index.html` and not `{outDir}/index.html`. This causes `#vinxi/spa/{router}` to break.

Generated base assets paths cause nothing to be loaded in build.

~Vite will look for index.html in project root.~ (can be solved by using the dev handler for non index.html)

~Vite's `transformIndexHtml` function will not transform the entry path unless `index.html` is present in project root.~ (solved: see dev handler for none index.html)
</details>

# Other Information

Closes #20

edivados commented on packages/vinxi/lib/router-modes.js:
```diff
@@ -368,7 +368,7 @@ const routerModes = {
 			invariant(router.mode === "spa", "Invalid router mode");
 			const appRoot = appConfig.root ?? process.cwd();
-			const root = router.root ?? appRoot;
+			const root = resolve.absolute(router.root, appRoot) ?? appRoot;
```
Requires some testing. It should also work with only `resolve.absolute(router.root, appRoot)`.

edivados commented:
Don't know why the dev multi spa test times out. The template itself works if started manually.

edivados commented:
> Don't know why the dev multi spa test times out. The template itself works if started manually. Must be some stupid mistake.

Hahah... dev fixture waits on the favicon and there was none.

edivados commented:
Set to draft because I have to do some more testing around handler being in a folder name like `react`. Rollup will fail to resolve it.

edivados commented on packages/vinxi/lib/router-modes.js:
```diff
@@ -354,11 +354,23 @@ const routerModes = {
+								html,
 							);
 
+							const { JSDOM } = await import("jsdom");
+							const jsdom = new JSDOM(transformedHtml);
+							const nodes = jsdom.window.document.querySelectorAll("[src],[href]");
+							nodes.forEach(node => {
+								const attr = node.hasAttribute("href") ? "href" : "src";
+								const path = node.getAttribute(attr) || "/";
+								if (!isAbsolute(path)) {
+									node.setAttribute(attr, join(viteDevServer.config.base, path));
+								}
+							});
+							transformedHtml = jsdom.serialize();
+
```
The same can be achived by temporarly generating a index.html and letting vite transform it.

```js
const indexPath = join(router.root, "index.html");
writeFileSync(indexPath, text);
let transformedHtml = await viteDevServer.transformIndexHtml(
  "/index.html",
  text,
  event.node.req.originalUrl
);
rmSync(indexPath);
```

edivados replied:
Writing to disk is not necessary. This seems to do the trick.
```js
await viteDevServer.transformIndexHtml(
  "/index.html",
  text,
  getRequestURL(event).href
);
```

edivados commented on packages/vinxi/lib/build.js:
```diff
@@ -313,8 +312,12 @@ export async function createBuild(app, buildConfig) {
  */
 async function createViteBuild(config) {
+	const cwd = process.cwd();
+	process.chdir(config.root);
 	const vite = await import("vite");
-	return await vite.build({ ...config, configFile: false });
+	const output = await vite.build({ ...config, configFile: false });
+	process.chdir(cwd);
+	return output;
```
To use relative paths for rollup working directory needs to be changed.

nksaraf commented:
Yeah this has been a tricky zone with how vite assumes things to be set up. I found that playing with the project root was one way to deal with this. But I love the experimentation here so we can reach a cleaner solution.

nksaraf commented:
Any blockers from my side?

edivados commented:
> Any blockers from my side?

Don't think so. I will have time to take another look at this next weekend. If you have any opinion or thoughts on the two commented changes let me know.

My opinion currently:

### Shared project root

Did not spend much time on this but feels like fighting Vite every step of the way. I think most problems here are related to asset paths and the transformIndexHtml issue also applies but this time for both handler types html and none html.

### transformIndexHtml using index.tsx

Would rather leave the transform up to Vite instead of doing something incomplete. I am intending to change it so the index.html is written to disk and deleted after the transform.


### Changing current working directory for Rollup

I think this makes sense when a router defines a separate root because it changes for example how Rollup resolves relative inputs `./solid/index.html` vs `./index.html`.

nksaraf commented:
Is there anything we need to do here? Or is this kinda done?

edivados commented:
This turned out to be a bit mor complicated than I expected at the time. At the end I could not figure out what the deal is with a shared root. The use case with different roots should work but would require more time to be looked through again and tested.

I will not have time to spend on something like this for a while. If it's ok with you we can close this.

edivados commented:
> This has to be figured out at some point, but might be better to throw errors for multiple SPA routers for now

Makes sense.

After reading what I wrote it sounds wrong... I am not asking to close the PR but am ok with it in case you decide to.

Right now I can't put in the time to try and solve the use case where multiple spa router share the same root dir. Better to not do it in this PR.

For my original use case where every spa router has a different root (isolated spas). There are only two issue left related to the build process.

- Public files not being copied
- Warning about out_dir not being in project root (strangely only for one router)

```
(!) outDir /workspaces/vinxi/test/.fixtures/multi-spa-prod-c7cut0n1tso/.vinxi/build/solid_entry is not inside project root and will not be emptied.
Use --emptyOutDir to override.
```

I updated the PR with the latest changes that I had locally and updated the tests to reflect the issues.
If I can solve these the PR is good to merge. No promises here for when I will get to it.

edivados commented on packages/vinxi/lib/build.js:
```diff
@@ -155,6 +155,7 @@ export async function createBuild(app, buildConfig) {
 							dir: router.dir,
 							baseURL: router.base,
 							passthrough: true,
+							fallthrough: true,
```
Not sure about the right default for all the router modes.

It seems nitro sets fallthrough to false under some conditions. For spa's this should be true otherwise 404 is returned for non existing paths instead of index.html.

Should this option be exposed on the router config? 
Should the public asset entry be generated for every spa router by default, removing the need to define it explicitly?

```js
{
  mode: "spa",
  publicDir: false
},
{
  mode: "spa",
  publicDir: {
    fallthrough: false
  }
}
```

nksaraf replied:
passthrough and fallthrough are different things ?? that's so confusing.

Can you explain which one does what? or is passthrough nothing lol.

edivados replied:
You are right I think `passthrough` isn't a thing... 🤣 can't find it in nitro or h3.
Ok so just have to replace it with `fallthrough` instead.

edivados commented on packages/vinxi/lib/build.js:
```diff
@@ -347,14 +352,17 @@ async function createRouterBuild(app, router) {
-					input: { handler: router.handler },
+					input: { handler: join(router.root, router.handler) },
 				},
 				target: "esnext",
 				outDir: join(router.outDir + "_entry"),
+				emptyOutDir: true,
 			},
```
Should be safe to always delete even when outDir is outisde of project root.
To be very safe it could only be set to true if outDir is inside app root.

nksaraf replied:
No its fine I like being strict about this. Causes too many stupid issues without this. Actually I manually delete all the folders before hand also but this fine too.

nksaraf commented:
Public files are not copied automatically. The whole project shares one top level public directory and the default vite one has been disabled.

Warnings are okay for now we can override them separately. 

If shared root doesn't work that's fine. Let's enable different roots first.

So if these are the conditions are we good to go ?

edivados commented:
> Public files are not copied automatically. The whole project shares one top level public directory and the default vite one has been disabled.
> 

Yeah.. I forgot to add static routers for the other spa. Maybe this could be improved int he future by generating them for spa routers with the possibility to opt out.

> Warnings are okay for now we can override them separately.
> 

Changed `emptyDir` to be always true when generating index.html. What are the odds that someone sets the outDir of a spa router to a directory outside the project root and accidentally delets something with a suffix of `_entry`. 
You can already do `outDir: "/"` and delete your entire project root on build which is worse. 🤣 

> If shared root doesn't work that's fine. Let's enable different roots first.
> 

Yes.

> So if these are the conditions are we good to go ?

Sure sounds good. Can you take a look at the last changes regarding the `fallthrough` option on the public assets. Otherwise I think this is done. The existing tests still pass.

https://github.com/nksaraf/vinxi/pull/21/files/6cbb0bacb2d59c3fc20b5554fafaa9ca736dd640#diff-cd0ed711204c4956e149d555193a953f5549ca0593f794659a518e41de740c81R171

https://github.com/nksaraf/vinxi/pull/21/files/6cbb0bacb2d59c3fc20b5554fafaa9ca736dd640#diff-cd0ed711204c4956e149d555193a953f5549ca0593f794659a518e41de740c81R158

nksaraf commented:
> You can already do `outDir: "/"` and delete your entire project root on build which is unfortunate.

I have half the minds to not allow to override outDir. We need the overrides to kind of be fixed for our setups to work. So its probably better than anybody who needs to move things around does it in a post:build hook. Us fs.cp and take it wherever you want.
